File: app/src/main/java/com/sumian/sleepdoctor/doctor/activity/PaymentActivity.java
package com.sumian.sleepdoctor.doctor.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.qmuiteam.qmui.widget.QMUIRadiusImageView;
import com.sumian.sleepdoctor.R;
import com.sumian.sleepdoctor.base.ActivityLauncher;
import com.sumian.sleepdoctor.base.BaseActivity;
import com.sumian.sleepdoctor.doctor.bean.DoctorService;
import com.sumian.sleepdoctor.doctor.bean.DoctorServicePackage;
import com.sumian.sleepdoctor.doctor.bean.DoctorServiceShopData;
import com.sumian.sleepdoctor.doctor.bean.PayOrder;
import com.sumian.sleepdoctor.doctor.contract.PayContract;
import com.sumian.sleepdoctor.doctor.dialog.PayDialog;
import com.sumian.sleepdoctor.doctor.presenter.PayPresenter;
import com.sumian.sleepdoctor.widget.TitleBar;
import com.sumian.sleepdoctor.widget.dialog.ActionLoadingDialog;
import com.sumian.sleepdoctor.widget.pay.PayCalculateItemView;
import com.sumian.sleepdoctor.widget.pay.PayItemGroupView;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * @author sm
 * on 2018/1/22.
 * desc:
 */

public class PaymentActivity extends BaseActivity<PayPresenter> implements View.OnClickListener, PayItemGroupView.OnSelectPayWayListener, TitleBar.OnBackClickListener, PayCalculateItemView.OnMoneyChangeCallback, PayContract.View {
    private static final String TAG = PaymentActivity.class.getSimpleName();
    private static final String ARGS_DOCTOR_SERVICE = "com.sumian.app.extra.doctor.service";
    private static final String ARGS_DOCTOR_SERVICE_PACKAGE_ID = "com.sumian.app.extra.doctor.service.packageId";
    @BindView(R.id.title_bar)
    TitleBar mTitleBar;

    @BindView(R.id.lay_group_icon)
    QMUIRadiusImageView mIvGroupIcon;

    @BindView(R.id.tv_desc)
    TextView mTvDesc;

    @BindView(R.id.pay_calculate_item_view)
    PayCalculateItemView mPayCalculateItemView;

    @BindView(R.id.tv_group_money)
    TextView mTvGroupMoney;

    @BindView(R.id.pay_group_view)
    PayItemGroupView mPayGroupView;

    @BindView(R.id.bt_pay)
    Button mBtPay;

    private String mPayChannel = "wx";

    private ActionLoadingDialog mActionLoadingDialog;

    private PayDialog mPayDialog;

    private DoctorService mDoctorService;
    private int mPackageId;
    private DoctorServicePackage mServicePackage;

    private static void startForResult(ActivityLauncher launcher, DoctorService doctorService, int packageId, int requestCode) {
        Intent intent = new Intent(launcher.getActivity(), PaymentActivity.class);
        intent.putExtra(ARGS_DOCTOR_SERVICE, doctorService);
        intent.putExtra(ARGS_DOCTOR_SERVICE_PACKAGE_ID, packageId);
        launcher.startActivityForResult(intent, requestCode);
    }

    public static void startForResult(ActivityLauncher launcher, DoctorService doctorService, int requestCode) {
        startForResult(launcher, doctorService, doctorService.getPackages().get(0).getId(), requestCode);
    }

    public static void startForResult(ActivityLauncher launcher, DoctorServiceShopData doctorServiceShopData, int requestCode) {
        startForResult(launcher, doctorServiceShopData.getService(), doctorServiceShopData.getPackageId(), requestCode);
    }

    @SuppressWarnings("unchecked")
    @Override
    protected boolean initBundle(Bundle bundle) {
        if (bundle != null) {
            this.mDoctorService = bundle.getParcelable(ARGS_DOCTOR_SERVICE);
            this.mPackageId = bundle.getInt(ARGS_DOCTOR_SERVICE_PACKAGE_ID);
            for (DoctorServicePackage servicePackage : mDoctorService.getPackages()) {
                if (servicePackage.getId() == mPackageId) {
                    mServicePackage = servicePackage;
                }
            }
        }
        return super.initBundle(bundle);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_main_shopping_car;
    }

    @Override
    protected void initWidget(View root) {
        super.initWidget(root);
        mTitleBar.setOnBackClickListener(this);
        mPayGroupView.setOnSelectPayWayListener(this);
        mPayCalculateItemView.setOnMoneyChangeCallback(this);
        mPayDialog = new PayDialog(root.getContext()).bindContentView(R.layout.dialog_pay);
        mPayDialog.setOwnerActivity(this);
        mPayDialog.bindPresenter(mPresenter);
    }

    @Override
    protected void initData() {
        super.initData();
        RequestOptions requestOptions = RequestOptions.placeholderOf(R.mipmap.ic_group_avatar).error(R.mipmap.ic_group_avatar);
        Glide.with(this).load(mDoctorService.getIcon()).apply(requestOptions).into(mIvGroupIcon);
        mTvDesc.setText(mDoctorService.getName());
        String priceText = getDoctorServicePackage().getPrice_text();
        SpannableString spannableString = new SpannableString(priceText);
        spannableString.setSpan(new ForegroundColorSpan(getResources().getColor(R.color.t4_color)), 0, priceText.indexOf("元"), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        mTvGroupMoney.setText(mDoctorService.getDescription());
        mPayCalculateItemView.setDefaultMoney(getDoctorServicePackage().getUnit_price());
    }

    @Override
    protected void initPresenter() {
        super.initPresenter();
        PayPresenter.Companion.init(this);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mPresenter.onPayActivityResultDelegate(requestCode, resultCode, data);
    }

    @Override
    @OnClick({R.id.bt_pay})
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.bt_pay:
                PayOrder payOrder = new PayOrder(mPayCalculateItemView.getCurrentMoney(), mPayChannel, "cny", mDoctorService.getName(), mDoctorService.getDescription(), getDoctorServicePackage().getId(), mPayCalculateItemView.getCurrentBuyCount());
                mPresenter.createPayOrder(this, payOrder);
                break;
            default:
                break;
        }
    }

    private DoctorServicePackage getDoctorServicePackage() {
        return mServicePackage;
    }

    @Override
    public void onSelectWechatPayWay() {
        this.mPayChannel = "wx";
    }

    @Override
    public void onSelectAlipayWay() {
        this.mPayChannel = "alipay";
    }

    @Override
    public void onBack(View v) {
        onBack();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        onBack();
    }

    private void onBack() {
        if (mActionLoadingDialog != null) {
            mActionLoadingDialog.dismiss();
        }
        finish();
    }

    @Override
    public void onMoneyChange(double money) {
        mBtPay.setEnabled(money > 0.00f);
        mBtPay.setAlpha(money > 0.00f ? 1.00f : 0.50f);
    }

    @Override
    public void setPresenter(PayContract.Presenter presenter) {
        this.mPresenter = (PayPresenter) presenter;
    }

    @Override
    public void onFailure(String error) {
        showToast(error);
        if (mActionLoadingDialog != null) {
            mActionLoadingDialog.dismiss();
        }
    }

    @Override
    public void onBegin() {
        mActionLoadingDialog = new ActionLoadingDialog().show(getSupportFragmentManager());
    }

    @Override
    public void onFinish() {
        if (mActionLoadingDialog != null) {
            mActionLoadingDialog.getDialog().cancel();
        }
    }

    @Override
    public void onCreatePayOrderSuccess() {
        mPresenter.doPay(this);
        showToast(R.string.create_order_success);
    }

    @Override
    public void onOrderPaySuccess(@NonNull String payMsg) {
        if (!mPayDialog.isShowing()) {
            mPayDialog.setPayStatus(PayDialog.PAY_SUCCESS).show();
        }
    }

    @Override
    public void onOrderPayFailed(@NonNull String payMsg) {
        showToast(payMsg);
        if (!mPayDialog.isShowing()) {
            mPayDialog.setPayStatus(PayDialog.PAY_FAILED).show();
        }
    }

    @Override
    public void onOrderPayInvalid(@NonNull String payMsg) {
        showToast(payMsg);
        if (!mPayDialog.isShowing()) {
            mPayDialog.setPayStatus(PayDialog.PAY_INVALID).show();
        }
    }

    @Override
    public void onOrderPayCancel(@NonNull String payMsg) {
        showToast(payMsg);
        if (mActionLoadingDialog != null) {
            mActionLoadingDialog.dismiss();
        }
    }

    @Override
    public void onCheckOrderPayIsOk() {
        if (mPayDialog != null && mPayDialog.isShowing()) {
            mPayDialog.cancel();
        }
        setResult(Activity.RESULT_OK);
        finish();
    }

    @Override
    public void onCheckOrderPayIsInvalid(@NonNull String invalidError) {
        onCheckOrderPayIsOk();
    }
}

File: client/src/main/java/com/sumian/sd/common/pay/activity/PaymentActivity.kt
package com.sumian.sd.common.pay.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import com.blankj.utilcode.util.ToastUtils
import com.sumian.common.base.BaseViewModelActivity
import com.sumian.common.image.ImageLoader
import com.sumian.common.statistic.StatUtil
import com.sumian.sd.R
import com.sumian.sd.buz.doctor.bean.DoctorService
import com.sumian.sd.buz.doctor.bean.DoctorServicePackage
import com.sumian.sd.buz.stat.StatConstants
import com.sumian.sd.common.pay.bean.PayCouponCode
import com.sumian.sd.common.pay.bean.PayOrder
import com.sumian.sd.common.pay.dialog.PayDialog
import com.sumian.sd.common.pay.presenter.PayPresenter
import com.sumian.sd.common.pay.widget.PayCalculateItemView
import com.sumian.sd.common.pay.widget.PayItemGroupView
import com.sumian.sd.widget.TitleBar
import com.sumian.sd.widget.dialog.ActionLoadingDialogV2
import kotlinx.android.synthetic.main.activity_main_shopping_car.*

/**
 * @author sm
 * on 2018/1/22.
 * desc:
 */

class PaymentActivity : BaseViewModelActivity<PayPresenter>(), View.OnClickListener, PayItemGroupView.OnSelectPayWayListener, TitleBar.OnBackClickListener, PayCalculateItemView.OnMoneyChangeCallback {

    companion object {

        private val TAG = PaymentActivity::class.java.simpleName

        private const val WECHAT_PAY_TYPE = "wx"
        private const val ALIPAY_PAY_TYPE = "alipay"

        private const val ARGS_DOCTOR_SERVICE = "com.sumian.app.extra.doctor.service"
        private const val ARGS_DOCTOR_SERVICE_PACKAGE_ID = "com.sumian.app.extra.doctor.service.packageId"

        @JvmStatic
        fun startForResult(activity: Activity, doctorService: DoctorService, packageId: Int, requestCode: Int) {
            val extras = Bundle()
            extras.putParcelable(ARGS_DOCTOR_SERVICE, doctorService)
            val intent = Intent(activity, PaymentActivity::class.java).apply {
                putExtras(extras)
                putExtra(ARGS_DOCTOR_SERVICE_PACKAGE_ID, packageId)
            }
            activity.startActivityForResult(intent, requestCode)
        }
    }

    private var mPayChannel = WECHAT_PAY_TYPE

    private val mActionLoadingDialog: ActionLoadingDialogV2  by lazy {
        ActionLoadingDialogV2(this)
    }

    private val mPayDialog: PayDialog  by lazy {
        val payDialog = PayDialog(this, object : PayDialog.Listener {
            override fun onRepayClick() {
                return pay()
            }
        }).bindContentView(R.layout.dialog_pay)
        payDialog.ownerActivity = this@PaymentActivity
        payDialog.bindPresenter(mViewModel!!)
        return@lazy payDialog
    }

    private var mDoctorService: DoctorService? = null

    private var mServicePackage: DoctorServicePackage? = null

    private var mPackage: DoctorServicePackage.ServicePackage? = null

    private var mIsCheckCouponCode = false
    private var mGoNextPay = false

    override fun initBundle(bundle: Bundle) {
        bundle.let {
            this.mDoctorService = bundle.getParcelable(ARGS_DOCTOR_SERVICE)
            val packageId = bundle.getInt(ARGS_DOCTOR_SERVICE_PACKAGE_ID)
            for (servicePackage in mDoctorService!!.service_packages) {
                if (servicePackage.id == packageId) {
                    this.mServicePackage = servicePackage
                    this.mPackage = mServicePackage!!.packages[0]
                    break
                }
            }
        }
    }

    override fun getLayoutId(): Int {
        return R.layout.activity_main_shopping_car
    }

    override fun initWidget() {
        super.initWidget()
        PayPresenter.init(this)
        nested_scroll_view.setOnClickListener {
            pay_calculate_item_view.closeKeyBoard()
        }
        title_bar.setOnBackClickListener(this)
        bt_pay.setOnClickListener(this)
        pay_group_view.setOnSelectPayWayListener(this)
        pay_calculate_item_view.setOnMoneyChangeCallback(this)
        StatUtil.event(StatConstants.page_pay)
    }

    override fun getPageName(): String {
        return StatConstants.page_pay
    }

    override fun initData() {
        super.initData()
        mDoctorService?.let {
            ImageLoader.loadImage(mDoctorService!!.icon, lay_group_icon)
            tv_name.text = mDoctorService?.name
            tv_desc.text = mServicePackage?.name
            pay_calculate_item_view.defaultMoney = mPackage?.unit_price!!.toDouble()
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        mViewModel?.onPayActivityResultDelegate(requestCode, resultCode, data)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.bt_pay -> {
                pay()
                StatUtil.event(StatConstants.click_pay)
            }
            else -> {
            }
        }
    }

    override fun onSelectWechatPayWay() {
        this.mPayChannel = WECHAT_PAY_TYPE
    }

    override fun onSelectAlipayWay() {
        this.mPayChannel = ALIPAY_PAY_TYPE
    }

    override fun onBack(v: View) {
        onBackPressed()
    }

    override fun onBackPressed() {
        dismissLoading()
        cancelPayDialog()
        finish()
    }

    override fun onMoneyChange(money: Double) {
        bt_pay.isEnabled = money > 0.00f
        bt_pay.alpha = if (money > 0.00f) 1.00f else 0.50f
    }

    override fun onCheckCouponCode(couponCode: String) {
        Log.e(TAG, "onCheckCouponCode: -------->")
        mGoNextPay = false
        checkCouponCode(false)
    }

    fun setPresenter(presenter: PayPresenter) {
        this.mViewModel = presenter
    }

    fun onFailure(error: String) {
        ToastUtils.showShort(error)
        dismissLoading()
    }

    fun onBegin() {
        showLoading()
    }

    fun onFinish() {
        dismissLoading()
    }

    fun onCreatePayOrderSuccess() {
        mViewModel?.doPay(this)
        ToastUtils.showShort(R.string.create_order_success)
    }

    fun onOrderPaySuccess(payMsg: String) {
        StatUtil.event(StatConstants.e_pay_success,
                mapOf(
                        "amount" to pay_calculate_item_view.currentMoney.toString(),
                        "quantity" to pay_calculate_item_view.currentBuyCount.toString(),
                        "channel" to if (mPayChannel == ALIPAY_PAY_TYPE) "支付宝" else "微信",
                        "productName" to mDoctorService!!.name
                ))
        if (!mPayDialog.isShowing) {
            mPayDialog.setPayStatus(PayDialog.PAY_SUCCESS).show()
        }
    }

    fun onOrderPayFailed(payMsg: String) {
        if (!mPayDialog.isShowing) {
            mPayDialog.setPayStatus(PayDialog.PAY_FAILED).show()
        }
    }

    fun onOrderPayInvalid(payMsg: String) {
        if (!mPayDialog.isShowing) {
            mPayDialog.setPayStatus(PayDialog.PAY_INVALID).show()
        }
    }

    fun onOrderPayCancel(payMsg: String) {
        if (!mPayDialog.isShowing) {
            mPayDialog.setPayStatus(PayDialog.PAY_CANCELED).show()
        }
        dismissLoading()
    }

    fun onCheckOrderPayIsOk() {
        cancelPayDialog()
        setResult(Activity.RESULT_OK)
        finish()
    }

    fun onCheckOrderPayIsInvalid(invalidError: String) {
        ToastUtils.showShort(invalidError)
        if (!mPayDialog.isShowing) {
            mPayDialog.setPayStatus(PayDialog.PAY_INVALID).show()
        } else {
            mPayDialog.setPayStatus(PayDialog.PAY_INVALID)
        }
    }

    fun onCheckOrderPayFinialIsInvalid(invalidError: String) {
        ToastUtils.showShort(invalidError)
        cancelPayDialog()
        setResult(Activity.RESULT_CANCELED)
        finish()
    }

    fun onCheckCouponCodeSuccess(payCouponCode: PayCouponCode?, payCouponCodeText: String, is2Pay: Boolean) {
        mIsCheckCouponCode = false
        mGoNextPay = false
        pay_calculate_item_view.updateCouponCodeTips(payCouponCode)
        if (is2Pay) {
            val payOrder = PayOrder(payCouponCodeText, null, null, null, pay_calculate_item_view.currentMoney, mPayChannel, "cny", mDoctorService!!.name, mDoctorService!!.description, null, mPackage!!.id, pay_calculate_item_view.currentBuyCount)
            mViewModel?.createPayOrder(this, payOrder)
        }
    }

    fun onCheckCouponCodeFailed(error: String, code: Int, payCouponCodeText: String?, is2Pay: Boolean) {
        mIsCheckCouponCode = false
        if (code == 1) {
            pay_calculate_item_view.updateCouponCodeTips(null)
            pay_calculate_item_view.updateCouponCodeFailed(error)
            ToastUtils.showShort(error)
        } else {
            ToastUtils.showShort(error)
            pay_calculate_item_view.updateCouponCodeTips(null)
        }
        if (code == 1 && is2Pay) {
            mGoNextPay = true
            ToastUtils.showShort(error)
        }
    }

    override fun showLoading() {
        if (mIsCheckCouponCode) {
            bt_pay.isEnabled = false
            return
        }
        mActionLoadingDialog.show()
    }

    override fun dismissLoading() {
        bt_pay.isEnabled = true
        if (mActionLoadingDialog.isShowing) {
            mActionLoadingDialog.dismiss()
        }
    }

    private fun pay() {
        Log.e(TAG, "go pay: -------->")
        if (mGoNextPay) {
            mIsCheckCouponCode = false
            val payOrder = PayOrder(null, null, null, null, pay_calculate_item_view.currentMoney, mPayChannel, "cny", mDoctorService!!.name, mDoctorService!!.description, null, mPackage!!.id, pay_calculate_item_view.currentBuyCount)
            mViewModel?.createPayOrder(this, payOrder)
        } else {
            checkCouponCode(true)
        }
    }

    private fun checkCouponCode(is2Pay: Boolean) {
        val payCouponCodeText = pay_calculate_item_view.getCouponCode()
        if (TextUtils.isEmpty(payCouponCodeText)) {
            mIsCheckCouponCode = false
            val payOrder = PayOrder(null, null, null, null, pay_calculate_item_view.currentMoney, mPayChannel, "cny", mDoctorService!!.name, mDoctorService!!.description, null, mPackage!!.id, pay_calculate_item_view.currentBuyCount)
            mViewModel?.createPayOrder(this, payOrder)
        } else {
            mIsCheckCouponCode = true
            mViewModel?.checkCouponCode(is2Pay, payCouponCodeText!!, mPackage!!.id)
        }
    }

    private fun cancelPayDialog() {
        mIsCheckCouponCode = false
        if (mPayDialog.isShowing) {
            mPayDialog.cancel()
        }
    }
}
